From a4bfbfe3454e0b995946e7eb5d9929989186cb04 Mon Sep 17 00:00:00 2001
From: toonn <toonn@toonn.io>
Date: Tue, 1 Feb 2022 15:06:41 +0100
Subject: [PATCH 2/3] Intercept old availability macros to use modern approach

---
 EXTERNAL_HEADERS/Availability.h | 80 +++++++++++++++++++++++++--------
 1 file changed, 61 insertions(+), 19 deletions(-)

diff --git a/EXTERNAL_HEADERS/Availability.h b/EXTERNAL_HEADERS/Availability.h
index f8f4463..6ccd4fd 100644
--- a/EXTERNAL_HEADERS/Availability.h
+++ b/EXTERNAL_HEADERS/Availability.h
@@ -187,25 +187,67 @@
 
 #include <AvailabilityInternal.h>
 
-#ifdef __IPHONE_OS_VERSION_MIN_REQUIRED
-    #define __OSX_AVAILABLE_STARTING(_osx, _ios) __AVAILABILITY_INTERNAL##_ios
-    #define __OSX_AVAILABLE_BUT_DEPRECATED(_osxIntro, _osxDep, _iosIntro, _iosDep) \
-                                                    __AVAILABILITY_INTERNAL##_iosIntro##_DEP##_iosDep
-    #define __OSX_AVAILABLE_BUT_DEPRECATED_MSG(_osxIntro, _osxDep, _iosIntro, _iosDep, _msg) \
-                                                    __AVAILABILITY_INTERNAL##_iosIntro##_DEP##_iosDep##_MSG(_msg)
-
-#elif defined(__MAC_OS_X_VERSION_MIN_REQUIRED)
-    #define __OSX_AVAILABLE_STARTING(_osx, _ios) __AVAILABILITY_INTERNAL##_osx
-    #define __OSX_AVAILABLE_BUT_DEPRECATED(_osxIntro, _osxDep, _iosIntro, _iosDep) \
-                                                    __AVAILABILITY_INTERNAL##_osxIntro##_DEP##_osxDep
-    #define __OSX_AVAILABLE_BUT_DEPRECATED_MSG(_osxIntro, _osxDep, _iosIntro, _iosDep, _msg) \
-                                                    __AVAILABILITY_INTERNAL##_osxIntro##_DEP##_osxDep##_MSG(_msg)
-
-#else
-    #define __OSX_AVAILABLE_STARTING(_osx, _ios)
-    #define __OSX_AVAILABLE_BUT_DEPRECATED(_osxIntro, _osxDep, _iosIntro, _iosDep)
-    #define __OSX_AVAILABLE_BUT_DEPRECATED_MSG(_osxIntro, _osxDep, _iosIntro, _iosDep, _msg)
-#endif
+#define API___IPHONE_2_0 2.0
+#define API___IPHONE_2_1 2.1
+#define API___IPHONE_2_2 2.2
+#define API___IPHONE_3_0 3.0
+#define API___IPHONE_3_1 3.1
+#define API___IPHONE_3_2 3.2
+#define API___IPHONE_4_0 4.0
+#define API___IPHONE_4_1 4.1
+#define API___IPHONE_4_2 4.2
+#define API___IPHONE_4_3 4.3
+#define API___IPHONE_5_0 5.0
+#define API___IPHONE_5_1 5.1
+#define API___IPHONE_6_0 6.0
+#define API___IPHONE_6_1 6.1
+#define API___IPHONE_7_0 7.0
+#define API___IPHONE_7_1 7.1
+#define API___IPHONE_8_0 8.0
+#define API___IPHONE_8_1 8.1
+#define API___IPHONE_8_2 8.2
+#define API___IPHONE_8_3 8.3
+#define API___IPHONE_8_4 8.4
+#define API___IPHONE_9_0 9.0
+#define API___IPHONE_9_1 9.1
+#define API___IPHONE_9_2 9.2
+#define API___IPHONE_9_3 9.3
+#define API___IPHONE_10_0 10.0
+#define API___IPHONE_10_1 10.1
+#define API___IPHONE_10_2 10.2
+#define API___IPHONE_10_3 10.3
+#define API___IPHONE_11_0 11.0
+#define API___IPHONE_11_1 11.1
+#define API___IPHONE_11_2 11.2
+#define API___IPHONE_11_3 11.3
+#define API___MAC_10_0 10.0
+#define API___MAC_10_1 10.1
+#define API___MAC_10_2 10.2
+#define API___MAC_10_3 10.3
+#define API___MAC_10_4 10.4
+#define API___MAC_10_5 10.5
+#define API___MAC_10_6 10.6
+#define API___MAC_10_7 10.7
+#define API___MAC_10_8 10.8
+#define API___MAC_10_9 10.9
+#define API___MAC_10_10 10.10
+#define API___MAC_10_10_2 10.10.2
+#define API___MAC_10_10_3 10.10.3
+#define API___MAC_10_11 10.11
+#define API___MAC_10_11_2 10.11.2
+#define API___MAC_10_11_3 10.11.3
+#define API___MAC_10_11_4 10.11.4
+#define API___MAC_10_12 10.12
+#define API___MAC_10_12_1 10.12.1
+#define API___MAC_10_12_2 10.12.2
+#define API___MAC_10_12_4 10.12.4
+#define API___MAC_10_13 10.13
+#define API___MAC_10_13_1 10.13.1
+#define API___MAC_10_13_2 10.13.2
+#define API___MAC_10_13_4 10.13.4
+#define __OSX_AVAILABLE_STARTING(_osx, _ios) __API_AVAILABLE(macos(API_##_osx),ios(API_##_ios))
+#define __OSX_AVAILABLE_BUT_DEPRECATED(_osxIntro, _osxDep, _iosIntro, _iosDep) __API_DEPRECATED("deprecated", macos(API_##_osxIntro,API_##_osxDep), ios(API_##_iosIntro,API_##_iosDep))
+#define __OSX_AVAILABLE_BUT_DEPRECATED_MSG(_osxIntro, _osxDep, _iosIntro, _iosDep, _msg) __API_DEPRECATED(_msg, macos(API_##_osxIntro,API_##_osxDep), ios(API_##_iosIntro,API_##_iosDep))
 
 
 #if defined(__has_feature)
-- 
2.17.2 (Apple Git-113)

